default/AI/ResearchAI.py

```diff
@@ -379,53 +394,141 @@ def generate_research_orders():
-            tuple(AIDependencies.DAMAGE_CONTROL_TECHS): partial(if_enemies, 0.1, 0.5),
-            tuple(AIDependencies.HULL_TECHS): get_hull_priority,
-            tuple(AIDependencies.ARMOR_TECHS): partial(ship_usefulness, if_enemies, MIL_IDX),
-            tuple(AIDependencies.ENGINE_TECHS): partial(ship_usefulness, partial(if_dict, choices, 'engine', true_val=0.6), None),
-            tuple(AIDependencies.FUEL_TECHS): partial(ship_usefulness, partial(if_dict, choices, 'fuel'), None),
-            tuple(AIDependencies.SHIELD_TECHS): partial(ship_usefulness, if_enemies, MIL_IDX),
-            tuple(AIDependencies.TROOP_POD_TECHS): partial(ship_usefulness, partial(if_enemies, 0.1, 0.3), TROOP_IDX),
-            tuple(AIDependencies.COLONY_POD_TECHS): partial(ship_usefulness, partial(const_priority, 0.5), COLONY_IDX),
-        })
-
-    # keys are individual full tech names
-    if not individual_priority_funcs:
-        individual_priority_funcs.update({
-            AIDependencies.PRO_MICROGRAV_MAN: partial(conditional_priority, partial(const_priority, 3.5), priority_low, None, ColonisationAI, 'got_ast'),
-            AIDependencies.PRO_ORBITAL_GEN: partial(conditional_priority, partial(const_priority, 3.0), priority_low, None, ColonisationAI, 'got_gg'),
-            AIDependencies.PRO_SINGULAR_GEN: partial(conditional_priority, partial(const_priority, 3.0), priority_low, partial(has_star, fo.starType.blackHole)),
-            AIDependencies.GRO_XENO_GENETICS: get_xeno_genetics_priority,
-            AIDependencies.LRN_XENOARCH: priority_low if foAI.foAIstate.aggression < fo.aggression.typical else partial(conditional_priority, partial(const_priority, 5.0), priority_low, None, ColonisationAI, 'gotRuins'),
-            AIDependencies.LRN_ART_BLACK_HOLE: get_artificial_black_hole_priority,
-            AIDependencies.GRO_GENOME_BANK: priority_low,
-            AIDependencies.CON_CONC_CAMP: partial(priority_zero),
-            AIDependencies.NEST_DOMESTICATION_TECH: priority_zero if foAI.foAIstate.aggression < fo.aggression.typical else partial(conditional_priority, partial(const_priority, 3.0), priority_low, None, ColonisationAI, 'got_nest'),
-        })
-
-    if not ALL_MISC_GROUP_TECH_NAMES:
-        ALL_MISC_GROUP_TECH_NAMES.extend(list(tech_name for misc_group in misc_group_priority_funcs for tech_name in misc_group))
-
-
-
-    report_adjustments = False
+    if not priority_funcs:
+        tech_handlers = (
+            (
+                AIDependencies.PRO_MICROGRAV_MAN,
+                partial(conditional_priority, partial(const_priority, 3.5), priority_low, None, ColonisationAI,
+                        'got_ast')
+            ),
+            (
+                AIDependencies.PRO_ORBITAL_GEN,
+                partial(conditional_priority, partial(const_priority, 3.0), priority_low, None, ColonisationAI,
+                        'got_gg')
+            ),
+            (
```

Comment:
spacing out the priority functions with extra newlines and other whitespace like this does help some to make the individual entries more easily readable (and would so more if we used a more consistent pattern of newlines and tabs akin to what we do in FOCS), but it doesn't seem to me to make a big benefit in that direction, and it makes it much more difficult to compare different ones to each other since you're unlikely to be able to see them on the screen at the same time.

I think I'd like to first explore using the approach in https://github.com/Cjkjvfnby/freeorion/pull/14 (and also applying it to const etc, not just conditional_priority), and see how much that improves readability with my original format.  

Converting the misc groups into individual tech entries in priority_funcs looks fine.

Also, we have so many references to AIDependencies.X, AIDependencies.Y, etc., I'm wondering if we should consider  either just doing a ```from AIDependencies import *``` (normally a bit risky, I know, but this seems a bit of a special case), or else simply at the top of this file doing a big set of simplifying definitions like
```
PRO_MICROGRAV_MAN = AIDependencies.PRO_MICROGRAV_MAN
PRO_ORBITAL_GEN = AIDependencies.PRO_ORBITAL_GEN
```